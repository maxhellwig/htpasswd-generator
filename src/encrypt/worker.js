import { encrypt } from "./encrypt";

export default global.onmessage = function(e) {
  // eslint-disable-next-line no-console
  console.log("Worker: Message received from main script");
  const [password, saltLength] = e.data;
  encrypt(password, saltLength);
};
