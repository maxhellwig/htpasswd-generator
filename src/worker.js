
import init, {hash_input} from "../bcrypt-wasm/pkg/bcrypt_wasm.js";

export default global.onmessage = function(event) {
    // eslint-disable-next-line no-console
    console.log("Worker: Message received from main script");
    const [password] = event.data;
    if (password) {
        init().then(() => {
           global.postMessage(hash_input(password));
        });
    }
};
