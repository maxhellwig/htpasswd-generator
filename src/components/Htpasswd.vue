<template>
  <div>
    <v-form v-on:submit="generatePassword" ref="form" v-model="valid">
      <v-card-text>
        <v-text-field
          label="Username"
          name="username"
          type="text"
          :rules="[v => !!v || 'Password is required']"
          required
          v-model.trim="username"
        >
        </v-text-field>
        <v-text-field
          label="Password"
          name="password"
          type="password"
          :rules="[v => !!v || 'Email is required']"
          required
          v-model.trim="password"
        >
        </v-text-field>
        <v-select label="Method" v-model="selectedMethod" :items="methods">
        </v-select>
        <v-text-field
          label="Salt length"
          type="number"
          id="salt"
          v-model.number="saltLength"
        >
        </v-text-field>
      </v-card-text>
      <v-card-actions>
        <v-btn
          id="generate-password"
          color="primary"
          :disabled="!valid"
          @click="generatePassword"
          >Generate</v-btn
        >
        <v-spacer></v-spacer>
        <v-btn @click="reset">Reset form</v-btn>
      </v-card-actions>
      <v-card-text>
        <template v-if="calculating">
          <code class="w-full p-4"
            ><v-progress-circular indeterminate color="primary"
          /></code>
        </template>
        <template v-else>
          <code class="w-full p-4"
            >{{ username }}{{ separator }}{{ result }}</code
          >
        </template>
      </v-card-text>
    </v-form>
  </div>
</template>

<script>
import Worker from "worker-loader!../encrypt/worker.js";
import { encrypt } from "../encrypt/encrypt";
let worker;
if (window.Worker) {
  worker = new Worker();
}
const METHODS = { BCRYPT: "bcrypt" };

export default {
  name: "Htpasswd",

  data() {
    return {
      valid: false,
      username: "",
      password: "",
      saltLength: 10,
      result: "",
      calculating: false,
      selectedMethod: METHODS.BCRYPT,
      methods: [METHODS.BCRYPT]
    };
  },
  computed: {
    separator() {
      return this.password !== "" ? ":" : "";
    }
  },
  methods: {
    reset() {
      this.username = "";
      this.password = "";
      this.result = "";
      this.calculating = false;
    },
    generatePassword() {
      if (this.valid) {
        this.calculating = true;
        this.result = "";
        if (window.Worker) {
          worker.postMessage([this.password, this.saltLength]);
        } else {
          encrypt(this.password, this.saltLength);
        }
      }
    }
  },
  mounted() {
    if (window.Worker) {
      worker.onmessage = ({ data }) => {
        this.calculating = false;
        this.result = data;
      };
    }
  }
};
</script>
<style>
.w-full {
  width: 100%;
}
.p-4 {
  padding: 1rem;
}
</style>
