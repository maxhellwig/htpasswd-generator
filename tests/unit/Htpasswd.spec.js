// Libraries
import Vuetify from "vuetify";

// Utilities
import { createLocalVue } from "@vue/test-utils";

const localVue = createLocalVue();

import Htpasswd from "../../src/components/Htpasswd";
import { shallowMount } from "@vue/test-utils";

import worker from "../../src/encrypt/worker";

jest.mock("../../src/encrypt/worker", () => {
  return {
    postMessage: jest.fn(),
    onmessage: jest.fn()
  };
});

describe("Htpasswd", () => {
  let vuetify;
  let wrapper;
  beforeEach(() => {
    vuetify = new Vuetify();
    wrapper = shallowMount(Htpasswd, {
      localVue,
      vuetify
    });
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("has name 'Htpasswd'", () => {
    expect(Htpasswd.name).toBe("Htpasswd");
  });
  it("renders as expected", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("resets inputs when calling reset()", () => {
    wrapper.setData({ password: "abc" });
    expect(wrapper.vm.password).toBe("abc");
    wrapper.vm.reset();
    expect(wrapper.vm.password).toBe("");
  });
  it("sends message to generate password to web worker", () => {
    const username = "abc";
    const password = "abc";
    const defaultSaltLength = 10;
    wrapper.setData({
      username: username,
      password: password,
      valid: true
    });
    const button = wrapper.find("#generate-password");
    button.trigger("click");
    expect(wrapper.vm.calculating).toBe(true);
    expect(worker.postMessage).toHaveBeenCalledWith([
      password,
      defaultSaltLength
    ]);
  });
});
