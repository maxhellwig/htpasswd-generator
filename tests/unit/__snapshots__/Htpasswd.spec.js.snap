// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Htpasswd has name 'Htpasswd' 1`] = `
<div>
  <v-form>
    <v-card-text>
      <v-text-field label="Username" name="username" type="text" rules="function (v) {
        /* istanbul ignore next */cov_1dasiliqer.f[4]++;
        cov_1dasiliqer.s[33]++;
        return (/* istanbul ignore next */(cov_1dasiliqer.b[6][0]++, !!v) || /* istanbul ignore next */(cov_1dasiliqer.b[6][1]++, 'Password is required')
        );
      }" required=""></v-text-field>
      <v-text-field label="Password" name="password" type="password" rules="function (v) {
        /* istanbul ignore next */cov_1dasiliqer.f[6]++;
        cov_1dasiliqer.s[35]++;
        return (/* istanbul ignore next */(cov_1dasiliqer.b[8][0]++, !!v) || /* istanbul ignore next */(cov_1dasiliqer.b[8][1]++, 'Email is required')
        );
      }" required=""></v-text-field>
      <v-select label="Method" items="bcrypt"></v-select>
      <v-text-field label="Salt length" type="number" id="salt"></v-text-field>
    </v-card-text>
    <v-card-actions>
      <v-btn color="primary" disabled="true">Generate</v-btn>
      <v-spacer></v-spacer>
      <v-btn>Reset form</v-btn>
    </v-card-actions>
    <v-card-text><code class="w-full p-4"></code></v-card-text>
  </v-form>
</div>
`;

exports[`Htpasswd renders as expected 1`] = `
<div>
  <v-form>
    <v-card-text>
      <v-text-field label="Username" name="username" type="text" rules="function (v) {
        /* istanbul ignore next */cov_1dasiliqer.f[4]++;
        cov_1dasiliqer.s[30]++;
        return (/* istanbul ignore next */(cov_1dasiliqer.b[5][0]++, !!v) || /* istanbul ignore next */(cov_1dasiliqer.b[5][1]++, 'Password is required')
        );
      }" required=""></v-text-field>
      <v-text-field label="Password" name="password" type="password" rules="function (v) {
        /* istanbul ignore next */cov_1dasiliqer.f[6]++;
        cov_1dasiliqer.s[32]++;
        return (/* istanbul ignore next */(cov_1dasiliqer.b[7][0]++, !!v) || /* istanbul ignore next */(cov_1dasiliqer.b[7][1]++, 'Email is required')
        );
      }" required=""></v-text-field>
      <v-select label="Method" items="bcrypt"></v-select>
      <v-text-field label="Salt length" type="number" id="salt"></v-text-field>
    </v-card-text>
    <v-card-actions>
      <v-btn id="generate-password" color="primary" disabled="true">Generate</v-btn>
      <v-spacer></v-spacer>
      <v-btn>Reset form</v-btn>
    </v-card-actions>
    <v-card-text><code class="w-full p-4"></code></v-card-text>
  </v-form>
</div>
`;
